drivers/fuel_gauge/sy24561/sy24561.c

```diff
@@ -0,0 +1,346 @@
+
+	LOG_DBG("RSOC: %" PRIu8 "%%", *soc_percent);
+	return 0;
+}
+
+static int sy24561_get_current_direction(const struct device *dev, uint16_t *current_direction)
+{
+	uint16_t status;
+
+	if (sy24561_read_reg(dev, SY24561_REG_STATUS, &status) != 0) {
+		LOG_ERR("Impossible to get status register");
+		return -ENODEV;
+	}
+
+	/* This comes from datasheet at page 6 */
+	*current_direction = IS_BIT_SET(status, 0);
+
+	return 0;
+}
+
+static int sy24561_get_version(const struct device *dev, uint16_t *version)
+{
+	if (sy24561_read_reg(dev, SY24561_REG_VERSION, version) != 0) {
+		LOG_ERR("Impossible to get version register");
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static int sy24561_get_config(const struct device *dev, uint16_t *config)
+{
+	if (sy24561_read_reg(dev, SY24561_REG_CONFIG, config) != 0) {
+		LOG_ERR("Impossible to get config register");
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static int sy24561_get_status(const struct device *dev, uint16_t *fg_status)
+{
+	uint16_t config;
+	int const ret = sy24561_get_config(dev, &config);
+
+	if (ret != 0) {
+		LOG_ERR("Failed to read config: %d", ret);
+		return ret;
+	}
+
+	/* This comes from datasheet at page 6 */
+	uint16_t const alarm = IS_BIT_SET(config, 5);
+
+	*fg_status = alarm;
+
+	return ret;
+}
+
+/** This function is to reset the alarm bit */
+static int sy24561_set_status(const struct device *dev, uint16_t const status)
+{
+	LOG_DBG("Setting status to %" PRIu16, status);
+
+	if (status != 0) {
+		LOG_ERR("Invalid status %" PRIu16 ", it should be 0", status);
+		return -EINVAL;
+	}
+
+	uint16_t config;
+	int const ret = sy24561_get_config(dev, &config);
+
+	if (ret != 0) {
+		LOG_ERR("Failed to read config: %d", ret);
+		return ret;
+	}
+
+	LOG_DBG("config register: 0x%x", config);
+
+	/* This comes from datasheet at page 6 */
+	uint16_t const status_mask = 1 << 5;
+
+	config &= ~status_mask;
+	LOG_DBG("new config register: 0x%x", config);
+
+	if (sy24561_write_reg(dev, SY24561_REG_CONFIG, config) != 0) {
+		LOG_ERR("Impossible to write config register");
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static int sy24561_set_alarm_threshold(const struct device *dev, uint16_t const percent_threshold)
+{
+	LOG_DBG("Setting SOC alarm threshold to %" PRIu16, percent_threshold);
+
+	uint16_t config;
+	int const ret = sy24561_get_config(dev, &config);
+
+	if (ret != 0) {
+		LOG_ERR("Failed to read config: %d", ret);
+		return ret;
+	}
+
+	LOG_DBG("config register: 0x%x", config);
+
+	/* This comes from datasheet at page 6 */
+	uint16_t const percent_threshold_max = 32;
+	uint16_t const percent_threshold_min = 1;
+
+	if (percent_threshold > percent_threshold_max) {
+		LOG_ERR("SOC alarm threshold %" PRIu16 " should be <= %" PRIu16, percent_threshold,
+			percent_threshold_max);
+		return -EINVAL;
+	}
+
+	if (percent_threshold < percent_threshold_min) {
+		LOG_ERR("SOC alarm threshold %" PRIu16 " should be >= %" PRIu16, percent_threshold,
+			percent_threshold_min);
+		return -EINVAL;
+	}
```

Comment:
If you are throwing an error here instead of changing it to a warning and correct the value reduce the portability.

Imagine this:  You use a different fuel-gauge which allows  to set an Alarm from 0-99%, then, in a later design/project etc., you want to use your application code again and you are using this fuel-gauge, which crash because this specific IC does not allow to set a threshold above 32%.  

As mentioned in the last review:  
> In this case it's better to throw a warning and compensate, instead of throwing an error

```suggestion
	if (percent_threshold > percent_threshold_max) {
		LOG_WRN("SOC alarm threshold %" PRIu16 " should be  <= %" PRIu16, percent_threshold,
			percent_threshold_max);
		percent_threshold = percent_threshold_max;;
	}

	if (percent_threshold < percent_threshold_min) {
		LOG_WRN("SOC alarm threshold %" PRIu16 " should be >= %" PRIu16, percent_threshold,
			percent_threshold_min);
		percent_threshold = percent_threshold_min;
	}
```